torchtune/modules/tokenizer.py

```diff
@@ -111,23 +127,36 @@ def tokenize_messages(
         for message in messages:
             # If assistant message, this is the end of a turn
             end_of_turn = message.role == "assistant"
+
             # Prepend BOS on start of new turns
             if start_of_turn:
                 tokenized_messages.append(self.bos_id)
                 mask.append(message.masked)
-                # Tokenize current message, append with masks
+
+            # We want to trim leading whitespace on the next message when
+            # (a) it is a continuation of the turn (i.e. not the first message)
+            # (b) the vocabulary explicitly encodes whitespace characters, and
+            # (c) the previous message did not end with a space
+            trim_leading_whitespace = (
+                (not start_of_turn)
+                and self.encodes_whitespace
+                and not prev_ends_with_space
```

Comment:
L153 as of this version. This is so that we can work with e.g. the grammar format

Reply:
this should throw an error because I don't see it defined on the very first message, it is only defined after. you should initialize it as False (or whichever is the default behavior) above the for loop

Reply:
Oh god I think it's breaking out of the and early in the test cases so we never hit this. Good catch